[sui-bcs-json] add a library for converting between bcs and json (and vice-versa) for transaction data

stefan-mysten opened the pull request:
## Description 
This PR adds support for a Rust implementation that exposes a C interface for supporting conversions between BCS and JSON formats for `TransactionData` type and its subtypes. More details are available in this issue: #13483.
Currently this is work in progress. 
## Test Plan 

Several tests are included. Plan to include also C tests. 

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### Release notes
This PR adds support for a Rust implementation that exposes a C interface for supporting conversions between BCS and JSON formats. More details are available in this issue: #13483.

amnn commented on crates/sui-bcs-json/src/lib.rs:
```diff
@@ -0,0 +1,388 @@
+
+/// Returns a BCS array (u8) from the input json string.
+#[ffi_export]
+fn json_to_bcs(type_name: char_p::Ref<'_>, json_data: char_p::Ref<'_>) -> repr_c::Vec<u8> {
```
C doesn't have namespacing so you will need to prefix these function names with something to make sure they don't conflict with other functions.  How about `sui_bcs_to_json` and `sui_bcs_from_json`?

amnn reviewed (commented):
This is a good start!  I left some initial comments on which types to give names to (I think you've got a good initial selection here), and thoughts on the initial FFI.

Type name wise, the main omission is handling `vector` (not your fault as we hadn't discussed this yet).  It won't be possible to handle it in the same way as the other types, because it could be nested arbitrarily deep, so you can't just map it to a call to `to_json::<T>` for some static `T`.  We can chat about the best way to do this.

Otherwise, I think you can be even more conservative with set of types that get names.  At this stage, a good test is to see which types the TypeScript SDK serializes with BCS -- those are probably the ones that need names for other SDKs to do the same.  (The primitive types also get names so that people can serialize/deserialize Move types as well).

An additional thought on the interface: When deserializing BCS, it may be worth exposing an interface where deserialization can partially consume the string being serialized, and communicate back how much of the buffer it used up.  This will allow people to serialize their custom types by successively serializing primitives.  E.g. an interface like so:

```
int sui_bcs_to_json(
    const char *type_name, 
    const uint8_t **bcs_buf,
    size_t len,
    const char **result,
);
```

- Made the return type `int` to return error codes.
- Namespaced the function to avoid name clashes.
- Accept the bcs input as a pointer-to-pointer, so that the function can update the location of the pointer on successful parse.
- Move the result to an out parameter, so that you are encouraged to check the return code first before inspecting the result.  If the operation failed, the result pointer can be set to `null` as well.

amnn commented on crates/sui-bcs-json/src/lib.rs:
```diff
@@ -0,0 +1,388 @@
+
+/// Frees a Rust-allocated string.
+#[ffi_export]
+fn rust_free_string(string: char_p::Box) {
```
following the naming scheme I mentioned elsewhere: `void sui_bcs_free_string(const char *)`.

stefan-mysten commented on crates/sui-bcs-json/src/bin/generate-headers.rs:
Right, I'll fix that.

stefan-mysten commented on crates/sui-bcs-json/src/lib.rs:
```diff
@@ -0,0 +1,388 @@
+}
+
+fn internal_bcs_to_json(typename: &str, data: Vec<u8>) -> Result<String, Error> {
+    let json = match typename {
```
Yes, this one is tricky.

bmwill commented on crates/sui-bcs-json/src/lib.rs:
```diff
@@ -0,0 +1,648 @@
+/// Deserializes the data to given type T, and then serializes it as a JSON value
+fn to_json<T: Serialize + for<'a> Deserialize<'a>>(data: &[u8]) -> Result<Value, Error> {
+    let bcs: T = bcs_from_bytes(data)?;
+    Ok(json!(bcs))
```
the json macro is generally used to construct json from literals, you may want to use [to_value](https://docs.rs/serde_json/latest/serde_json/fn.to_value.html) instead because using this macro also swallows the error turning it into a panic.

This comment applies to most usages of this macro throughout this file

bmwill commented on crates/sui-bcs-json/src/lib.rs:
```diff
@@ -0,0 +1,648 @@
+    match typename {
+        "u8" => {
+            let bcs: u8 = bcs_from_bytes(data)?;
+            Ok(json!(bcs))
+        }
+        "u16" => {
+            let bcs: u16 = bcs_from_bytes(data)?;
+            Ok(json!(bcs))
+        }
+        "u32" => {
+            let bcs: u32 = bcs_from_bytes(data)?;
+            Ok(json!(bcs))
```
why are these not converted to strings?

stefan-mysten replied:
In the TS reimplementation of the BCS library, it was decided to only keep `u64,u128,u256` as strings, thus I thought we should be consistent across our implementations.

bmwill commented on crates/sui-bcs-json/src/lib.rs:
```diff
@@ -0,0 +1,648 @@
+///
+/// By default, u8, u16, u32 are serialized as JSON numbers,
+/// while u64, u128, and U256 are serialized as JSON strings.
+fn string_integer_to_json(typename: &str, data: &[u8]) -> Result<Value, Error> {
```
do we want to use actual types here vs stringly matching things?

bmwill commented on crates/sui-bcs-json/src/lib.rs:
```diff
@@ -0,0 +1,648 @@
+            assert!(vec.len() == vec.capacity());
+            let ptr = vec.as_mut_ptr();
+            let len = vec.len();
+            std::mem::forget(vec); // prevent deallocation in Rust
```
You could also follow this [example](https://doc.rust-lang.org/stable/std/vec/struct.Vec.html#examples-3) and use ManuallyDrop.

bmwill commented on crates/sui-bcs-json/src/lib.rs:
```diff
@@ -0,0 +1,648 @@
+    let data = unsafe { std::slice::from_raw_parts(ptr[0] as *const u8, len) };
+    match internal_bcs_to_json(type_name.to_str(), &data, pretty) {
+        Ok(res) => {
+            let cstr = std::ffi::CString::new(res).unwrap().into_raw();
```
You probably don't want to unwrap here and should handle the error

bmwill commented on crates/sui-bcs-json/src/lib.rs:
```diff
@@ -0,0 +1,648 @@
+fn sui_bcs_json_free_array(ptr: *mut u8, len: *mut size_t) {
+    let len = len as usize;
+    unsafe {
+        drop(Vec::from_raw_parts(ptr, len, len));
+    }
+}
+
+/// Frees a Rust-allocated string.
+#[ffi_export]
+unsafe fn sui_bcs_json_free_string(pointer: *const safer_ffi::c_char) {
+    drop(std::ffi::CString::from_raw(pointer as *mut i8));
```
you don't need to call drop, you just need to build the thing and the compiler will drop it for you

bmwill reviewed (commented):
One thing you'll want to be careful of is that order matters in BCS but in json the order may or may not be preserved, based on the type.

stefan-mysten commented on crates/sui-bcs-json/Cargo.toml:
```diff
@@ -0,0 +1,33 @@
+serde_json = "1.0"
+bcs = "0.1.4"
+anyhow = "1.0"
+safer-ffi = "0.1"
```
That's fair. It made some things simpler to start prototyping, but I will try to deprecate it.

stefan-mysten replied:
Deprecated

stefan-mysten commented:
@bmwill Lots of updates here 
* got rid of `safer_ffi`
* added more tests
* finished the vector implementation
* added more docs

Still needs to be decided: which types to support directly, and how to support custom structs?

stefan-mysten commented:
@bmwill - would love to hear your feedback on the changes here. 
Some updates
* got rid of `safer_ffi`
* added vector implementation (thanks @amnn for helping here)
* added some docs + tests
* 
Still to be decided: 
* which types to support (now we have a few that I will remove) or how to support custom structs
* how to package this for external consumption

stefan-mysten commented on crates/sui-bcs-json/Cargo.toml:
```diff
@@ -0,0 +1,26 @@
+[workspace]
+
+[dependencies]
+sui-types = {path = "../sui-types"}
```
We'll have to deal with these before we merge it. For now, it makes it quicker to compile & test.

stefan-mysten replied:
Sorry, what I meant was to import them via `workspace`.

stefan-mysten commented on crates/sui-bcs-json/src/lib.rs:
```diff
@@ -0,0 +1,841 @@
+
+/// Deserializes the data to given type T, and then serializes it as a JSON value
+fn to_json<T: Serialize + for<'a> Deserialize<'a>>(data: &mut &[u8]) -> Result<Value> {
+    let bcs: T = bcs::from_reader_no_check(data)?;
```
This is a function that I added in my own `bcs` fork. The reason is that the original code https://github.com/zefchain/bcs/blob/main/src/de.rs#L89 calls `.end()?` which ensures that the BCS has the specified length and nothing else, which kinda defeats the purpose of being a reader.

bmwill replied:
Please use the upstream bcs library and not a fork.

> which ensures that the BCS has the specified length and nothing else, which kinda defeats the purpose of being a reader.

This does not defeat the purpose and in fact is *required* for correctness. BCS is not a self-describing format and as such in order to use BCS properly you need to know beforehand the exact type of a serialized payload before you begin to deserialize it and you must fully and completely consume the entire serialized payload when deserializing to ensure that your local understanding of the type layout is correct otherwise you violate the determinism and canonical principles of the format. For example if there is a type T where two entities disagree about the layout we can run into problems:

```rust
struct Ta {
    field1: u64,
    field2: u8,
}

struct Tb {
    field1: u64,
    field2: u8,
    field3: u8,
}
```

If entity B serializes type T and sends it to entity A, if A doesn't exhaust the whole payload it'll end up with an incomplete view of the message that it was sent. This is a huge issue when it comes to cryptography and if this T was a transaction then entity A and B would disagree on the hash of the transaction. In general we need the serialization to fully round-trip everywhere.

```
payload == ser(des(payload))
```

amnn replied:
We were not planning on forking `bcs` but @stefan-mysten needed to do this to get things working, and we wanted to get your thoughts on what the best path is towards getting something that works for this case, and is palatable to include in upstream `bcs`.

Agreed that we cannot in general treat deserialization as "done" if there is data left in a payload, but the use case that this is trying to cater to is one where the C library does not know what the schema of the type is either, and is acting as a conduit for the host language (which does know) where the schema is known (e.g. for a Move type from some arbitrary package).

stefan-mysten replied:
As per our last discussion, reverted this change and fork deleted :D

stefan-mysten commented on crates/sui-bcs-json/src/lib.rs:
```diff
@@ -0,0 +1,841 @@
+        .split(&['<', '>'][..])
+        .collect::<Vec<_>>()
+        .get(1)
+        .unwrap();
```
need to fix this.

bmwill commented on crates/sui-bcs-json/src/lib.rs:
```diff
@@ -0,0 +1,841 @@
+        typename
+    };
+    let bcs_bytes = match type_ {
+        "u8" | "u16" | "u32" | "u64" | "u128" | "u256" => string_integer_to_bcs(typename, data),
+        "bool" => to_bcs::<bool>(data),
+        "vector" => to_bcs_from_vector(typename, data),
+        "Argument" => to_bcs::<Argument>(data),
+        "CallArg" => to_bcs::<CallArg>(data),
+        "ObjectDigest" => to_bcs::<ObjectDigest>(data),
+        "ObjectID" => to_bcs::<ObjectID>(data),
+        "ProgrammableTransaction" => to_bcs::<ProgrammableTransaction>(data),
+        "SequenceNumber" => to_bcs::<SequenceNumber>(data),
+        "SuiAddress" => to_bcs::<SuiAddress>(data),
+        "String" => to_bcs::<String>(data),
+        "TransactionData" => to_bcs::<TransactionData>(data),
+        "TransactionKind" => to_bcs::<TransactionKind>(data),
```
I'm not sure i fully understand how this library is intended to be used. Given the stated goal in the README i expect this is intended as a mechanism for other languages which don't have bcs support (serde-reflection provides bcs support for a number of languages though) the ability to write and craft a transaction in JSON and then have this tool take that and then write out its canonical BCS format. Given that i'm not sure why this tool would support all of these types and not just the high level "Transaction" type or other top-level sui data types. I would not expect anyone to create a transaction by going back and forth multiple times into this library but instead would expect it to be a more one-shot style where you have a fully formed txn.

amnn replied:
> write and craft a transaction in JSON and then have this tool take that and then write out its canonical BCS format.

This is mostly true, but there are some edge cases that require the introduction of finer grained types:

- Serialization of `Pure` transaction arguments, which requires the the ability serialise primitive types and vectors of those.
- The ability to ser/de arbitrary Move types, which would be composed of those primitives -- we briefly entertained the idea of supporting a DSL for describing such types so that you could call into this library once with a DSL describing the schema (the extensions we are planning for Object Display require a similar DSL which we were thinking of borrowing) -- but I thought it would be less jarring to offer an API that allows the client to chain calls into the serializer/deserializer for these cases.
- I suggested that @stefan-mysten look through the types that the TS SDK serializes and deserializes to BCS to get an indication of the kinds of calls an SDK in another language might make.

Some of the types that are mentioned here may not fit into any of these cases and we can get rid of those though (`Argument`, `CallArg`, `ProgrammableTransaction` come to mind).

stefan-mysten replied:
Thanks both for your comments. I have not removed those types before asking for review, which I understand now that it created a lot of confusion. Here's what TS BCS supports: https://sui-typescript-docs.vercel.app/typescript/bcs and here https://sui-typescript-docs.vercel.app/bcs#description

amnn replied:
Note that there is a slight difference between the list of supported types by TS BCS that you've linked here and the list that I meant.  We only need to support the "root" types that actually get passed to `bcs.ser` or `bcs.se` in the TS SDK.  The list above includes the names of all intermediate types as well, which we don't need here, because that mapping is handled within Rust by serde derivations.

stefan-mysten replied:
Right!

bmwill reviewed (changes requested):
There are a couple of things that I believe need to be tweaked for correctness (eg using the bcs format properly) and I also think that it would be worthwhile to discuss what the actual user journey for using this library would look like as i'm probably failing to understand the expected flow. Understanding the expected use cases could allow us to improve the safety and ergonomics of this library.

amnn commented on crates/sui-bcs-json/src/lib.rs:
```diff
@@ -0,0 +1,744 @@
+        "String" => to_json::<String>(data),
+        "TransactionData" => to_json::<TransactionData>(data),
+        "TransactionKind" => to_json::<TransactionKind>(data),
+        x if x.starts_with("vector") => to_json_from_vector(typename, data),
```
...this was a stale comment from before -- feel free to ignore!

stefan-mysten commented on crates/sui-bcs-json/Cargo.toml:
```diff
@@ -0,0 +1,26 @@
+serde = "1.0"
+serde_json = "1.0"
+anyhow = "1.0"
+bcs = { git = "https://github.com/stefan-mysten/bcs"}
```
This was just for experiments, I plan to delete the fork once we figure out how to use the upstream code / add something to that codebase.

stefan-mysten replied:
Forked deleted!

stefan-mysten commented on crates/sui-bcs-json/Cargo.toml:
```diff
@@ -0,0 +1,26 @@
+[lib]
+crate-type = ["cdylib", "staticlib", "rlib"]
+
+[workspace]
```
Will update it. For now, it was like this for build speed.

stefan-mysten commented on crates/sui-bcs-json/src/lib.rs:
```diff
@@ -281,11 +350,7 @@ fn internal_bcs_from_json(typename: &str, data: &str) -> Result<Vec<u8>> {
-        .collect::<Vec<_>>()
-        .get(1)
-        .unwrap();
+    let typename = is_vector(typename).unwrap();
```
this is safe to unwrap because typename is first checked if it is a vector before being passed to this function.

stefan-mysten commented on crates/sui-bcs-json/src/lib.rs:
```diff
@@ -146,75 +147,143 @@ fn to_bcs<T: Serialize + for<'a> Deserialize<'a>>(data: &str) -> Result<Vec<u8>>
-        "u64" => {
-            let bcs: u64 = bcs::from_bytes(data)?;
-            Ok(serde_json::to_value(bcs.to_string())?)
+impl serde::Serialize for MyValue {
```
The compiler complained that there is no impl of `Serialize` for `MyValue`, thus I added one.

stefan-mysten commented:
@amnn - this is ready for re-review. A few things changed since last time:
* support for `vector` through this TypeLayout pattern that we discussed
* added types that are used in the BCS TS SDK: `MultiSig / MultiSigPublicKey / SuiAddress / String`. 
* primitive types are usually lowercase, Sui / Move specific types are uppercase.

Will add these days some tests that are shared between this library and the BCS TS SDK.

amnn commented on crates/sui-bcs-json/README.md:
```diff
@@ -0,0 +1,71 @@
+
+# Design choices
+
+The library was designed to be as simple as possible to allow users to convert to JSON or to BCS by passing the `typename` (as BCS requires to know before hand the type of the encoded data). Specifically, the types that are supported to be converted as of now are the `TransactionKind` and its subtypes from the main `Sui Transaction` type [transaction.rs](https://github.com/MystenLabs/sui/blob/mai ...
```
```suggestion
The library was designed to be as simple as possible to allow users to convert to JSON or to BCS by passing the `typename` (as BCS is schema-less so requires type information to be supplied during deserialization). Specifically, the types that are supported to be converted as of now are the `TransactionKind` and its subtypes from the main `Sui Transaction` type [transaction.rs](https://github.com/MystenLabs/sui/blob/main/crates/sui-types/src/transaction.rs)
```

amnn replied:
The types that are supported may need to be tweaked?

amnn commented on crates/sui-bcs-json/README.md:
````diff
@@ -0,0 +1,71 @@
+The library offers two functions for serializing and deserializing data:
+* **sui_bcs_from_json**
+```c
+size_t sui_bcs_from_json(const char *type_name,
````
nit: I have a slight preference for calling this `sui_json_to_bcs` to make the distinction between this API and the next one more obvious.

amnn commented on crates/sui-bcs-json/README.md:
```diff
@@ -0,0 +1,71 @@
+
+
+## Memory management
+There are two functions that should be used to free Rust allocated data (`Vec<u8>`` and `String`):
```
nit: This should be given in terms of the C types, i..e `const uint8_t *` and `const char *`.

amnn commented on crates/sui-bcs-json/README.md:
````diff
@@ -0,0 +1,71 @@
+## Memory management
+There are two functions that should be used to free Rust allocated data (`Vec<u8>`` and `String`):
+
+```c
````
We should consider merging this into one freeing function, because both types here are basically the same thing.  This would require us being able to always recover the length of the allocation from a pointer to it (similar to how `malloc` and `free` work).

stefan-mysten replied:
That makes sense, fixed!

amnn commented on crates/sui-bcs-json/README.md:
````diff
@@ -0,0 +1,71 @@
+```
+
+## Errors
+The library uses two return error codes to indicate an error occurred: 1 or 2. A return code of 0 indicates that the function call was successful. 
````
Can you document what these error codes mean?

amnn commented on crates/sui-bcs-json/sui_bcs_json_bindings.h:
```diff
@@ -0,0 +1,74 @@
+ * by calling the sui_bcs_json_free_array function.
+ *
+ * # Safety
+ * Unsafe function.
```
I'm not sure what this `Safety` section is meant to convey.

stefan-mysten replied:
Clippy complains without a # Safety comment. I think I can add a ignore lint for it though, but that was my first thought to do to get rid of the issue.

amnn commented on crates/sui-bcs-json/examples/main.c:
```diff
@@ -0,0 +1,64 @@
+    if (result_0 == 1) {
+        printf("Error code is %d. ", result_0);
+        int buffer = sui_last_error_length();
+        char error_msg [buffer];
```
I would personally `malloc` this rather than store it on the stack, given we don't know how big this is going to be (and then `free` it after the `printf`).  Mostly for the sake of setting a good example for people using this as a reference.

amnn commented on crates/sui-bcs-json/src/lib.rs:
```diff
@@ -0,0 +1,829 @@
+            let mut vec = output;
+            vec.shrink_to_fit();
+            assert!(vec.len() == vec.capacity());
+            let ptr = vec.as_mut_ptr();
+            let len = vec.len();
+            // Prevent running `vec`'s destructor so we are in complete control
+            // of the allocation.
+            let _ = ManuallyDrop::new(vec);
+            unsafe {
+                *result = ptr;
+                *length = len;
+            }
+            0
```
```suggestion
            let slice = vec.leak();
            unsafe {
                *result = slice.as_ptr();
                *length = slice.len();
            }
            0
```

stefan-mysten replied:
Cool!

amnn commented on crates/sui-bcs-json/src/lib.rs:
```diff
@@ -0,0 +1,829 @@
+}
+
+/// Deserializes the data to given type T, and then serializes it as a JSON value
+fn to_json<T: Serialize + for<'a> Deserialize<'a>>(data: &mut &[u8]) -> Result<Value> {
```
Do you still need the existential lifetime qualification, or can you get rid of it now?  Also, can you get rid of the `&mut` now?

```suggestion
fn to_json<T: Serialize + Deserialize>(data: &[u8]) -> Result<Value> {
```

stefan-mysten replied:
If we go with the updated ser/de, these do not make sense anymore.

amnn commented on crates/sui-bcs-json/src/lib.rs:
```diff
@@ -0,0 +1,829 @@
+}
+
+/// Deserializes the string data to a given type T, and then serializes it as a BCS array
+fn to_bcs<T: Serialize + for<'a> Deserialize<'a>>(data: &str) -> Result<Vec<u8>> {
```
ditto for the existential lifetime here:

```suggestion
fn to_bcs<T: Serialize + Deserialize>(data: &str) -> Result<Vec<u8>> {
```

stefan-mysten replied:
If we go with the updated ser/de, these do not make sense anymore.

amnn commented on crates/sui-bcs-json/src/lib.rs:
```diff
@@ -0,0 +1,829 @@
+    let bcs: Result<T, _> = serde_json::from_str(data);
+    match bcs {
+        Ok(bcs) => Ok(bcs::to_bytes(&bcs)?),
+        Err(_) => to_bcs_from_string(data),
```
What's the rationale for this fallback logic?  I'm a little uneasy about having a pattern like: "try this, and if it fails, try this other thing" in a library like this, but maybe I don't understand why it's here, can you explain in a comment? Or maybe we can get rid of it altogether.

stefan-mysten replied:
Does not matter anymore, but I think the rationale was that this was a way to support encoding numbers as strings. I disliked this a lot too, so we can get rid of it.

amnn reviewed (commented):
Leaving some surrounding comments, let me come back to this PR for the core code that uses type layout etc.

amnn commented:
I decided to do a clean room implementation of the core serialisation/deserialisation logic, using serde's `Serialize` and `DeserializeSeed`.  Here it is:

https://gist.github.com/amnn/cb02dfd0a6cd53fd64fe108669d1bbee

It's very similar to what you have here, but here are the main differences -- I think they largely boil down to your current implementation mixing some of the patterns from your old implementation with the new:

- The `Value` type that acts as the intermediary needs to contain a variant for every type (not just the primitives, but also `TransactionKind`, etc).
- We are missing the `address` primitive.
- We don't need a tree-like type for `TypeLayout` -- just the string representation of the type suffices.
- There is a generic pattern (of recursively calling `serialize` or `deserialize`) that can be applied naturally no matter what types you want to support.

amnn commented:
Another important difference:  The gist uses `DeserializeSeed` to go in both directions (from BCS to JSON and from JSON to BCS), while I think your current implementation does not use it for the latter.

One issue that writing the example highlighted though is consistently stringifying numbers: `serde_json` expects to see a number literal when parsing a numeric type (it will complain with a type error if it sees a string instead), so even if you special case how primitives are serialized and deserialized (which can be done) you will still face a problem when trying to deserialize a type like `TransactionKind` -- all its internal integers will still come out as JSON number literals.

stefan-mysten commented:
>  even if you special case how primitives are serialized and deserialized (which can be done) you will still face a problem when trying to deserialize a type like `TransactionKind` -- all its internal integers will still come out as JSON number literals.

That's a good catch!

stefan-mysten commented:
> Another important difference: The gist uses `DeserializeSeed` to go in both directions (from BCS to JSON and from JSON to BCS), while I think your current implementation does not use it for the latter.
> 
> One issue that writing the example highlighted though is consistently stringifying numbers: `serde_json` expects to see a number literal when parsing a numeric type (it will complain with a type error if it sees a string instead), so even if you special case how primitives are serialized and deserialized (which can be done) you will still face a problem when trying to deserialize a type like `TransactionKind` -- all its internal integers will still come out as JSON number literals.

So then, do we want to keep parity with TS BCS, where `u64`, `u128`, `u256` are serialized as strings?

amnn commented:
> So then, do we want to keep parity with TS BCS, where u64, u128, u256 are serialized as strings?

I'd like to hear your thoughts on this, and @hayes-mysten as well, but here are the things that I tried (none of which panned out):

- Annotating the integer values in more complex types using `serde_as(as = "DisplayFromStr")`.  The problem with this approach is that it also affects BCS serde, which we don't want.
- Deserializing the BCS into a JSON `Value`, and then replacing all the integers with strings in that before converting it into a string.  This also affects `u8`, `u16` and `u32`, and it means then when we convert from JSON to BCS, we can no longer distinguish between a stringified number and a string that happens to contain a number.
- Creating a custom `Serializer`/`Deserializer` pair that redirects serialization of u64, u128, etc to first stringify, and vice versa on the way back.  This plan almost has legs, but (a) it's a lot of work (Looking at the implementations of `Serializer` and `Deserializer` from `serde_json`), and (b) there's an open question of how it would interpret `u256` (although maybe this isn't an issue in practice, and that's already serialised and deserialised as a string).

stefan-mysten commented:
My thoughts on this are more long-term thinking in terms of risk having to deal with unforeseen problems.
What if we do not serialize them as strings, and someone somehow uses it together with the TS BCS, then we might encounter incompatibilities and errors. 

From another point of view, as we discussed, I think both the TS BCS and this library should share a test suite, and thus, their design choices with regards on how to serialize/deserialize `u64, u128, and u256` should match.

amnn commented:
> I'm not a rust expert, but I did find this snippet which seems to handle differentiating specific Number types.

The trouble with this approach is that it's deciding whether to convert to a string or not based on the value, rather than the type in the schema (i.e. if the type is `u64` but the value is `42`, then it will stay as a numeric literal), and as you mentioned:

> This doesn't cover the second half of the issue of converting from JSON to BCS. I don't have a good grasp of how this currently works, but my you need the BCS schema to do this conversion, which should mean that we know the target bcs types, and just need the parser to know how to Go from Value:Number | Value:String to u64

This is where the pain really starts, because these schemas are all defined in the core codebase, by deriving functions to serialize and deserialize them, using codegen.  My last bullet point suggests the way around this -- a JSON deserializer that sees that we're looking for a u64 or u128 and looks instead for a string.  This will 10x the implementation size of this project though, which is why I was not keen to go down that road.

A question for you both: We know that JavaScript's JSON parser will not handle large numeric literals, but how likely is it that someone would generate JSON from this library and feed it into a JS application?  This is intended as a tool for languages that don't have native BCS support, and JS/TS is not one of those.

If the answer to this is "unlikely", then should we revaluate the goal that the TS implementation and this implementation have parity?  For me the answer seems to be based on whether other languages' JSON parsers treat big numbers appropriately or not (for example `serde_json` with the `arbitrary_precision` feature will store numbers as strings to avoid this issue, similarly Python has big ints).

bmwill commented:
@stefan-mysten is it time? Are we reviving this effort?

stefan-mysten commented:
@bmwill yes, I think we should start looking at it again. 

I basically migrated to the types in the `sui-rust-sdk` repo. I have a few question, but need to poke a bit more to understand a few things: previous MultiSig type - there's now MultiSigCommittee, and a few others, so I need to understand the differences and what it's needed.
Also, we used to have TransactionData and V1 of that but that seems to be different now in the `sui-rust-sdk`. Need to read the code a bit to see what's new there.

One thing I observed is that `u256` is private. Let me know when you think we could make it public.

Hopefully I can poke a bit more today and then I can ask some questions! 
Let me know if you have other thoughts or suggestions on reviving this.

mitsosf commented:
Hello @stefan-mysten, I'm making a SUI SDK for PHP and BCS has been slowing me down the most 🥲 

I'm tempted to make a Rust lib wrapper directly, but C would def be better supported. Is this PR something you or the project are still actively pursuing?

stefan-mysten commented:
Hi @mitsosf, yes are still planning to get this finished but we don't have an exact date yet! 

What's your timeline?

mitsosf commented:
@stefan-mysten 
No hard timeline/deadline really on my end, it's more of a hobby project for now. I've implemented wallets and most of the read functionality for the SDK and wanted to move on to sending transactions.

Are you aware of any other way I can input JSON - or sth similar - and get back the serialized payload using Rust or any other SKDs I could wrap?

stefan-mysten commented:
You **might** be able to do so with the `pySui` SDK for Python: https://pysui.readthedocs.io/en/latest/index.html
but I am not 100% sure.

stefan-mysten commented:
I'll see if I can make some time to look again into this project, but can't make any promises yet. Please feel free to ping again in this thread in a few weeks if nothing seems to be moving 😅

stefan-mysten commented:
@mitsosf have a look at this community project: https://github.com/VAR-META-Tech/Rust2C-Sui-SDK
Maybe it can help you in the meantime.
